corehq/apps/app_manager/models.py

```diff
@@ -4175,36 +4176,36 @@ def jad_settings(self):
         return settings
 
     def create_jadjar(self, save=False):
-        try:
-            return (
-                self.lazy_fetch_attachment('CommCare.jad'),
-                self.lazy_fetch_attachment('CommCare.jar'),
-            )
-        except (ResourceError, KeyError):
-            built_on = datetime.datetime.utcnow()
-            all_files = self.create_all_files()
-            jad_settings = {
-                'Released-on': built_on.strftime("%Y-%b-%d %H:%M"),
-            }
-            jad_settings.update(self.jad_settings)
-            jadjar = self.get_jadjar().pack(all_files, jad_settings)
-            if save:
-                self.built_on = built_on
-                self.built_with = BuildRecord(
-                    version=jadjar.version,
-                    build_number=jadjar.build_number,
-                    signed=jadjar.signed,
-                    datetime=built_on,
+        with CriticalSection(['create_jadjar_' + self._id]):
+            try:
+                return (
+                    self.lazy_fetch_attachment('CommCare.jad'),
+                    self.lazy_fetch_attachment('CommCare.jar'),
                 )
+            except (ResourceError, KeyError):
+                built_on = datetime.datetime.utcnow()
+                all_files = self.create_all_files()
+                jad_settings = {
+                    'Released-on': built_on.strftime("%Y-%b-%d %H:%M"),
+                }
+                jad_settings.update(self.jad_settings)
+                jadjar = self.get_jadjar().pack(all_files, jad_settings)
+                if save:
+                    self.built_with = BuildRecord(
+                        version=jadjar.version,
+                        build_number=jadjar.build_number,
+                        signed=jadjar.signed,
+                        datetime=built_on,
+                    )
 
-                self.lazy_put_attachment(jadjar.jad, 'CommCare.jad')
-                self.lazy_put_attachment(jadjar.jar, 'CommCare.jar')
+                    self.lazy_put_attachment(jadjar.jad, 'CommCare.jad')
+                    self.lazy_put_attachment(jadjar.jar, 'CommCare.jar')
 
-                for filepath in all_files:
-                    self.lazy_put_attachment(all_files[filepath],
-                                             'files/%s' % filepath)
+                    for filepath in all_files:
+                        self.lazy_put_attachment(all_files[filepath],
+                                                 'files/%s' % filepath)
```

Comment:
Hey @nickpell sorry to catch this so late, but I just noticed the way this is done won't result in the change we were trying to make. Also just looking at it, I think installing on ODK will no longer work, but maybe I'm missing how that ends up happening. (Did you test that that works?)

This function (`create_jadjar`) does a lot more than just create the `CommCare.jad` and `CommCare.jar` attachments. This line in particular attaches all build files to the build, so right now —with no `'files/profile.ccpr'`, `files/suite.xml`, etc.—`make_build` isn't making a build at all, just the shell of one with no output files. (I imagine that's why you had to add `create_jadjar` back to all of the tests.)

So what we really want to do instead (and it might be easiest to start over) is rename `create_jadjar` to `create_build_files` or something and then remove

```python
jad_settings = {
    'Released-on': built_on.strftime("%Y-%b-%d %H:%M"),
}
jad_settings.update(self.jad_settings)
jadjar = self.get_jadjar().pack(all_files, jad_settings)
# ...
self.lazy_put_attachment(jadjar.jad, 'CommCare.jad')
self.lazy_put_attachment(jadjar.jar, 'CommCare.jar')
```

from this function and put them in another function called `create_jadjar_from_build_files` that you call on a built app. That function should loop through its attachments and fetch the ones prefixed with `files/` to get `all_files` and pull `built_on` out of `self.built_with`

```python
all_files = {
    filename[:-len('files/')]: self.lazy_fetch_attachment(filename)
    for filename in self._attachments if filename.startswith('files/')
}
built_on = self.built_with.datetime
```

and then do the things that we're removing from the (now-named) `create_build_files` function.

Let me know if you want to talk it through Wednesday.